### ISSUE TITLE
Support Sphinx `:doc:` and `:ref:` by replacement

### ISSUE THREAD

**Issue by TibsAtWork** [jdkato/txtlint#470]
Note: as this has to do with URLs, I have read https://github.com/errata-ai/vale/issues/320

We are using vale 2.20.0

Sphinx provides two special roles for cross referencing between documents, `:doc:` and `:ref:`, which are commonly used, and can cause problems when checking documents with Vale.

For instance
```
  See :doc:`this reference <sibling>`.
```
will provide a link to the `sibling` document in the same directory.

Since Vale uses `docutils` to process the reStructuredText files, those directives are not recognised. Instead, an error report is put into the output HTML. Vale then checks the content of that error report. Unfortunately, the reference `sibling` will also be in the text that is checked, which can lead to spelling errors.

A work-around for this would be to replace each occurrence of
```
  :ROLE:`this reference <REFERENCE>`
```
(where `ROLE` is `:doc:` or `:ref:`) with
```
  `this reference <REFERENCE>`__
```
(two underscores at the end to stop `docutils` complaining about possible duplicates) before passing the text to `docutils`.

While `REFERENCE` is unlikely to be a valid URL, that doesn't matter at this stage, as we're not actually trying to follow or validate the actual URLs.

I believe this is consistent with the current fix for the `glossary` directive in https://github.com/errata-ai/vale/blob/v2/internal/lint/rst.go

I can produce a PR to demonstrate this, if you wish.

Documentation for these two roles is at:
* https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html?highlight=roles#role-doc
* https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html?highlight=roles#role-ref

**Comment by jdkato** [comment]
> Since Vale uses docutils to process the reStructuredText files, those directives are not recognised. Instead, an error report is put into the output HTML. Vale then checks the content of that error report. Unfortunately, the reference sibling will also be in the text that is checked, which can lead to spelling errors.

Can you provide an example of this happening? The expected result is that `rst2html` wraps the text in a `problematic` class, which is then ignored by Vale:

```html
<p>See <a href="#id1"><span class="problematic" id="id2">:doc:`this reference &lt;sibling&gt;`</span></a>.</p>
```

There should be no error messages in the output.

**Comment by TibsAtWork** [comment]
Apologies for not giving an example earlier - I should have done so.

Given a `.vale.ini` containing:
```
StylesPath = "styles"

[*.rst]
BasedOnStyles = test
```
and a `styles/test/substitutions.yml` containing
```
extends: substitution
message: "Use '%s' instead of '%s'."
ignorecase: true
level: error
swap:
  timeseries: time-series
```
and the reStructuredText file `test.rst`
```
A test
======

`An inline link <choosing-timeseries-database.html>`_

:doc:`A doc link <choosing-timeseries-database>`
```
then:
```
$ vale --version
vale version 2.20.1

$ vale test.rst

 test.rst
 6:28  error  Use 'time-series' instead of    test.substitutions
              'timeseries'.

✖ 1 error, 0 warnings and 0 suggestions in 1 file.
```
My local `rst2html` produces
```
<p><a class="reference external" href="choosing-timeseries-database.html">An inline link</a></p>
<p><a href="#system-message-1"><span class="problematic" id="problematic-1">:doc:`A doc link &lt;choosing-timeseries-database&gt;`</span></a></p>
<div class="system-message" id="system-message-1">
<p class="system-message-title">System Message: ERROR/3 (<tt class="docutils">test.rst</tt>, line 6); <em><a href="#problematic-1">backlink</a></em></p>
Unknown interpreted text role &quot;doc&quot;.</div>
</div>
```
which sounds the same as you indicate - that is, wrapping it in a `problematic` class.

Of course, ideally I'd like the `A doc link` text to be checked just as the `An inline link` is for an inline link.

**Comment by jdkato** [comment]
I can't reproduce this: using the exact configuration you've provided, I get the expected output of no alerts.

The `rst2html` output you've shown also looks correct (although Vale uses [a few flags](https://github.com/errata-ai/vale/blob/v2/internal/lint/rst.go#L32)), but could you provide the following information just in case?

```console
$ rst2html.py --version
rst2html.py (Docutils 0.17.1 [release], Python 3.9.13, on darwin)
```

**Comment by TibsAtWork** [comment]
It does look like a problem with my configuration.

Locally, I have
```
$ rst2html.py --version
rst2html.py (Docutils 0.17.1 [release], Python 3.10.1, on darwin)
```
and also
```
$ rst2html --version
rst2html (Docutils 0.19, Python 3.10.6, on darwin)
```
so I don't *think* that should matter.

However, we just (4 days ago) upgraded our vale-action workflow to `uses: errata-ai/vale-action@reviewdog` instead of the previous `uses: errata-ai/vale-action@master`, and it looks as if the problem has gone away in our CI.
Specifically, that `test.rst` file no longer gives an error, as it would have done before, and it is now possible to use `:doc:` where we had been having to work around with docutils-style inline links.

I shall continue to investigate what the problem is with my command line setup, but it sounds like a configuration issue at my end, so this issue can be closed if you wish.

Thank you very much for the time you've taken to look at this.
